src/pty.rs

```diff
@@ -0,0 +1,72 @@
+use std::ffi::CStr;
+use std::os::unix::io::RawFd;
+
+use libc;
+
+use {Errno, Error, fcntl, Result};
+
+pub fn grantpt(fd: RawFd) -> Result<()> {
+    if unsafe { libc::grantpt(fd) } < 0 {
+        return Err(Error::last().into());
+    }
+
+    Ok(())
+}
+
+pub fn posix_openpt(flags: fcntl::OFlag) -> Result<RawFd> {
+    let fd = unsafe {
+        libc::posix_openpt(flags.bits())
+    };
+
+    if fd < 0 {
+        return Err(Error::last().into());
+    }
+
+    Ok(fd)
+}
+
+pub fn ptsname(fd: RawFd) -> Result<String> {
+    let name_ptr = unsafe { libc::ptsname(fd) };
+    if name_ptr.is_null() {
+        return Err(Errno::ENOTTY.into());
+    }
+
+    let name_cstr = unsafe { CStr::from_ptr(name_ptr) };
+    let name = match name_cstr.to_str() {
```

Comment:
It is probably safe to ignore utf8 validation errors as we get the string from the operating system:
see https://github.com/nix-rust/nix/blob/master/src/sys/utsname.rs#L65